pkg/costmodel/aggregation.go

```diff
@@ -0,0 +1,2004 @@
+		disableCache = true
+	}
+
+	// attempt to retrieve cost data from cache
+	var costData map[string]*CostData
+	var err error
+	cacheData, found := a.CostDataCache.Get(key)
+	if found && !disableCache && !noCache {
+		ok := false
+		costData, ok = cacheData.(map[string]*CostData)
+		cacheMessage = fmt.Sprintf("ComputeAggregateCostModel: L1 cache miss: %s, L2 cost data cache hit: %s", aggKey, key)
+		if !ok {
+			log.Errorf("ComputeAggregateCostModel: caching error: failed to cast cost data to struct: %s", key)
+		}
+	} else {
+		log.Infof("ComputeAggregateCostModel: missed cache: %s (found %t, disableCache %t, noCache %t)", key, found, disableCache, noCache)
+
+		costData, err = a.Model.ComputeCostDataRange(promClient, a.CloudProvider, window, resolution, "", "")
+		if err != nil {
+			if prom.IsErrorCollection(err) {
+				return nil, "", err
+			}
+			if pce, ok := err.(prom.CommError); ok {
+				return nil, "", pce
+			}
+			if strings.Contains(err.Error(), "data is empty") {
+				return nil, "", &EmptyDataError{err: err, window: window}
+			}
+			return nil, "", err
+		}
+
+		// compute length of the time series in the cost data and only compute
+		// aggregates and cache if the length is sufficiently high
+		costDataLen := costDataTimeSeriesLength(costData)
+
+		if window.Hours() < 1.0 {
+			// scale hourly cost data down to fractional hour
+			costData = ScaleHourlyCostData(costData, resolution.Hours())
+		}
+
+		if costDataLen == 0 {
+			return nil, "", &EmptyDataError{window: window}
+		}
+		if costDataLen >= minCostDataLength && !noCache {
+			log.Infof("ComputeAggregateCostModel: setting L2 cache: %s", key)
+			a.CostDataCache.Set(key, costData, cacheExpiry)
+		}
+	}
+
+	c, err := a.CloudProvider.GetConfig()
+	if err != nil {
+		return nil, "", err
+	}
+	discount, err := ParsePercentString(c.Discount)
+	if err != nil {
+		return nil, "", err
+	}
+	customDiscount, err := ParsePercentString(c.NegotiatedDiscount)
+	if err != nil {
+		return nil, "", err
+	}
+
+	sc := make(map[string]*SharedCostInfo)
+	if !disableSharedOverhead {
+		for key, val := range c.SharedCosts {
+			cost, err := strconv.ParseFloat(val, 64)
+			durationCoefficient := window.Hours() / util.HoursPerMonth
+			if err != nil {
+				return nil, "", fmt.Errorf("unable to parse shared cost %s: %s", val, err)
+			}
+			sc[key] = &SharedCostInfo{
+				Name: key,
+				Cost: cost * durationCoefficient,
+			}
+		}
+	}
+
+	idleCoefficients := make(map[string]float64)
+	if allocateIdle {
+		duration, offset := window.ToDurationOffset()
+
+		idleDurationCalcHours := window.Hours()
+		if window.Hours() < 1 {
+			idleDurationCalcHours = 1
+		}
+		duration = fmt.Sprintf("%dh", int(idleDurationCalcHours))
+
+		if a.ThanosClient != nil {
+			offset = thanos.Offset()
+			log.Infof("ComputeAggregateCostModel: setting offset to %s", offset)
+		}
+
+		idleCoefficients, err = a.ComputeIdleCoefficient(costData, promClient, a.CloudProvider, discount, customDiscount, duration, offset)
```

Comment:
@AjayTripathy this logic is 1:1, but looks weird to me (shouldn't we be careful to not overwrite an offset that was already assigned? don't we need to also adjust the start proportionally?) so maybe we can review this.